### ISSUE TITLE
Missing trait items false positive

### ISSUE THREAD

**Issue by alygin** [intellij-rust/intellij-rust#1009]
[Specialization](https://github.com/rust-lang/rfcs/pull/1210/) makes the current implementation of "Missing trait items" annotation prone to false positives:

```rust
#[derive(Clone)]
struct Foo<T> { t: T }

trait Bar<T> {
    fn bar(t: T);
}

impl<T> Bar<T> for Foo<T> {
    default fn bar(t: T) {}
}

impl<T: Clone> Bar<T> for Foo<T> {}
//^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  Not all trait items implemented, missing `bar`
```

Just noticed it, didn't think of what would be the better way to deal with it yet.

**Comment by alygin** [comment]
I think the current state of the resolve engine doesn't allow us to implement an absolutely correct specialization analysis and tell for sure which trait items are implemented for a struct and which are missing. Please, correct me if I'm wrong.

So, I would go with a simplified solution that is prone to false negatives, but removes the mentioned false positives. Particularly, I have two solutions in mind:

1. Don't annotate `impl` blocks that have type parameters polybounds (like `impl<T: Clone> ...` or `impl<T> ... where T: Send` etc) since they are potential specializations. This solution is simple and lightweight but switches off the annotation for many implementations that are not real specializations.

2. Extend the solution 1 by finding all the implementations of the given trait for the given struct and check if there're `default` implementations of the missing items from the `impl` block in question. Remove what we found from the list of missing items without additional checks. But I'm actually not sure that such solution won't bring some new kind of false positives, and it's pretty heavy (though we can do the analysis only when there're really missing items and the `impl` block is a specialization candidate as described above).

@matklad, any thoughts on this?

**Comment by matklad** [comment]
2 seems to be a reasonable approach. 

>we can do the analysis only when there're really missing items

I think this will cover absolute majority of cases, so I wouldn't be afraid about performance. And we need to make impl search fast anyway :)

**Comment by matklad** [comment]
But another option is just to wait until someone reports a false positive bug report: I think that perhaps specialization is not that abundant in practice? Though I don't have any real numbers.

**Comment by alygin** [comment]
It might not be abundant but it can be found in the real life. For instance, [here's an example](https://github.com/rust-lang/rust/blob/be760566cf938d11d34c2f6bd90d8fd0f67c2344/src/libcollections/vec.rs#L1697) of this false positive in the `Vec` from the stdlib (the preceding `impl` block contains the default implementation of `from_iter()`).

**Comment by abonander** [comment]
I've been adding specialization under a `nightly` feature in my crates and ran into this. In my case the specialization is an optimization for a special case that could impact real-world performance.

This impl overrides only one item out of four in its parent impl:
https://github.com/abonander/img_hash/blob/43cfba59fd548d4afbac0113a87722812b14bdef/src/lib.rs#L645-L653

Parent:
https://github.com/abonander/img_hash/blob/43cfba59fd548d4afbac0113a87722812b14bdef/src/lib.rs#L610-L623
